### PR TITLE
[SYCL-MLIR]: Uniformity analysis (part3)

### PR DESCRIPTION
**Pull request by etiotto**
This patch extends the uniformity analysis to allow it to determine the uniformity of loop induction variables and loop carried SSA values.

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp [hunk 1]
```diff
@@ -134,13 +133,79 @@ void UniformityAnalysis::visitOperation(
...
+
+  // If the lattice on any operand isn't yet initialized, bail out.
+  if (anyOfUniformityIsUninitialized(op->getOperands())) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "UA: Operand(s) uniformity not yet initialized\n");
+    return;
+  }
+
+  // Infer the uniformity of the loop IV by analyzing the loop bounds and step.
+  if (auto loop = dyn_cast<LoopLikeOpInterface>(op)) {
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:157]: what do you think of creating a visit function for loop, and returns optional Uniformity.
if return value is std::nullopt, then `return SparseDataFlowAnalysis::visitNonControlFlowArguments(op, successor, argLattices, firstIndex);`, else `return propagateAllIfChanged(*iv, *result);`?

**Review by sommerlukas — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp [hunk 2]
```diff
@@ -134,13 +133,79 @@ void UniformityAnalysis::visitOperation(
...
+      // case them here. Note that the step is always a constant.
+      ValueRange lbOperands = affineLoop.getLowerBoundOperands();
+      ValueRange ubOperands = affineLoop.getUpperBoundOperands();
+      boundsAndStep.append(lbOperands.begin(), lbOperands.end());
+      boundsAndStep.append(ubOperands.begin(), ubOperands.end());
+    } else {
+      std::optional<OpFoldResult> lb = loop.getSingleLowerBound();
+      std::optional<OpFoldResult> ub = loop.getSingleUpperBound();
+      std::optional<OpFoldResult> st = loop.getSingleStep();
+
```
Comment (jopperm) [inline polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:177]: Quite a bit of optionality here -- can you maybe add a comment to explain in which situation the `OpFoldResult` is there, but does not contain a `Value`?

File: polygeist/test/polygeist-opt/uniformity.mlir [hunk 3]
```diff
@@ -35,25 +35,84 @@ func.func @test1(%arg0 : i1, %arg1: memref<?x!sycl_nd_item_2>)  {
...
+
+  // COM: The result yielded by a loop with uniform bounds and step and a uniform 
+  // COM: loop carried value is uniform.
+  // CHECK: test1_v6, uniformity: uniform
+  %v6 = scf.for %ii = %c0 to %c4 step %c1 iter_args(%sum = %c0) -> index {
+    %next_sum = arith.addi %ii, %sum : index
+    scf.yield %next_sum : index
+  } {tag = "test1_v6"}
+
+  // COM: If a uniform loop carried value is added to a non-uniform value the resulted yielded
```
Comment (jopperm) [inline polygeist/test/polygeist-opt/uniformity.mlir:87]: Typo: ... the result yielded ...

**Review by jopperm — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp (hunk 2 above)
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:177]: Sure. Just answering your question here but I will add a comment to the code as well. 
`LoopLikeOpInterface::getSingleLowerBound` delegates its implementation to operations that implements the `LoopLikeOpInterface`.  Some operations might return a `std::nullopt` and some will return a `OpFoldResult` (like `scf::ForOp`). An `OpFoldResult` is `PointerUnion<Attribute, Value>` so it may be a constant (i.e `Attribute`) or a `Value`. 

In summary `getSingleLowerBound`, `getSingleUpperBound` may:
   - return `nullopt` if the operation has no lower bound to give (for whatever reason), or
   - return a `PointerUnion<Attribute, Value>` which is going to be:
       + a Attribute if, for example, the loop bound is known a compile time, or
        + a Value otherwise.

File: polygeist/test/polygeist-opt/uniformity.mlir (hunk 3 above)
Comment (etiotto) [reply polygeist/test/polygeist-opt/uniformity.mlir:87]: fixed

File: polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp (hunk 1 above)
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp:157]: done

File: polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp (hunk 2 above)
Comment (jopperm) [reply polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp]: Thanks, got it!

**Review by whitneywhtsang — APPROVED** [review]
